### PR TITLE
refactor(compiler-cli): better error messages when external strings u…

### PR DESCRIPTION
**Pull request by pmvald**
…sed for template and styles in local compilation mode

In local compilation mode it is not possible to use an imported string for component's template or styles as it cannot be resolved statically in compile time. There are some such use cases in g3 and potentially devs might incorporate such pattern. At the moment such pattern will cause the local compilation fail with generic error messages (e.g., so and so at position 1 is not a reference, etc). This change makes specific error messages with helpful hints for such cases. These new error messages can help devs to quickly resolve the issue as well as make it possible to identify existing issues in g3.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by devversion — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts [hunk 1]
```diff
@@ -461,6 +461,47 @@ function extractQueriesFromDecorator(
...
+  if (compilationMode === CompilationMode.LOCAL) {
+    if (Array.isArray(value)) {
+      for (const entry of value) {
+        if (entry instanceof DynamicValue && entry.isFromUnknownIdentifier()) {
+          const relatedInformation = traceDynamicValue(expression, entry);
+
+          const chain: ts.DiagnosticMessageChain = {
+            messageText: `Unknown identifier used as styles string: ${
+                entry.node
+                    .getText()} (did you import this string from another file? This is not allowed in local compilation mode. Please either inline it or move it to an .css file)`,
```
[the remark this answers is not in the document]
Comment (pmvald) [reply packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts]: Done!

File: packages/compiler-cli/src/ngtsc/annotations/component/src/resources.ts [hunk 2]
```diff
@@ -148,6 +149,27 @@ export function extractTemplate(
...
+      // cannot be resolved in single file mode. We warn the user here about the situation
+      // explicitly.
+      if (compilationMode === CompilationMode.LOCAL && resolvedTemplate instanceof DynamicValue &&
+          resolvedTemplate.isFromUnknownIdentifier()) {
+        const relatedInformation = traceDynamicValue(template.expression, resolvedTemplate);
+
+        const chain: ts.DiagnosticMessageChain = {
+          messageText: `Unknown identifier used as template string: ${
+              template.expression
+                  .getText()} (did you import this string from another file? This is not allowed in local compilation mode. Please either inline it or move it to an .html file)`,
```
[the remark this answers is not in the document]
Comment (pmvald) [reply packages/compiler-cli/src/ngtsc/annotations/component/src/resources.ts]: Done!

File: packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts [hunk 3]
```diff
@@ -461,6 +461,47 @@ function extractQueriesFromDecorator(
...
 }
 
+export function parseDirectiveStyles(
+    directive: Map<string, ts.Expression>, evaluator: PartialEvaluator,
+    compilationMode: CompilationMode): null|string[] {
+  if (!directive.has('styles')) {
+    return null;
+  }
+
+  const expression = directive.get('styles')!;
```
[the remark this answers is not in the document]
Comment (pmvald) [reply packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts]: Done!

File: packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts [hunk 4]
```diff
@@ -461,6 +461,48 @@ function extractQueriesFromDecorator(
...
+
+  if (!expression) {
+    return null;
+  }
+
+  const value = evaluator.evaluate(expression);
+
+  // Create specific error if any string is imported from external file in local compilation mode
+  if (compilationMode === CompilationMode.LOCAL) {
+    if (Array.isArray(value)) {
```
Comment (crisbeto) [inline packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts:477]: nit: maybe move the `isArray` check one level above the reduce the amount of nesting?

**Review by crisbeto — APPROVED** [review]
LGTM

File: packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts (hunk 4 above)
Comment (pmvald) [reply packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts]: Done!

**Review by devversion — APPROVED** [review]

**Review by crisbeto — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 0b901a814bbac82ea12b6ff612d4fd5c1e403ea9.

**Pull request closed without merging**